### ISSUE TITLE
[Question] Why does the Ensemble Selection implementation not return the best ensemble found during the greedy search? 

### ISSUE THREAD

**Issue by LennartPurucker** [automl/auto-sklearn#1477]
# Short Question Description
Why does the implementation of Ensemble Selection not return the best ensemble found during the greedy search? 

Following the algorithm presented in Caruana et al. (2004), the final step (step 4) of the algorithm should be to select the best ensemble found during all of the the n iterations from the greedy search in the previous step. 
However, the current implementation selects the last found/evaluated ensemble by default, that is, the full list of selected base models. 

I would like know whether this was done on purpose or was ignored, for example, because of its potentially marginal impact. 

I have came across the difference to the original algorithm while comparing the Ensemble Selection Implementation from [AutoGluon](https://github.com/awslabs/autogluon/blob/master/core/src/autogluon/core/models/greedy_ensemble/ensemble_selection.py#L174), a [re-implementation of the original paper](https://github.com/dclambert/pyensemble/blob/master/ensemble.py#L509),  and [AutoSklearn ](https://github.com/automl/auto-sklearn/blob/master/autosklearn/ensembles/ensemble_selection.py#L191).

# System Details
* As far as I am aware, no version of auto-sklearn has a different implementation of ensemble selection. 

# References
* [Caruana et al. (2004)](https://dl.acm.org/doi/pdf/10.1145/1015330.1015432)

**Comment by eddiebergman** [comment]
Hey @LennartPurucker,

Happy to see your around these parts :)

I believe the last ensemble found during one instance of ensemble building is "almost" guaranteed to be the best. Every iteration, the model that improves predictions the most will be added to the ensemble, including any model that is already present in the ensemble, in which case it is just given more weight. 

The negated view of this is a bit more illuminating, if the last ensemble produced is not the best, then at some point, predictions were added that degraded the performance of the ensemble. I'm not sure fully but I guess it could be possible that the ensemble could reach a state where the best model to add could slightly degrade performance but as you mentioned, its a "potentially marginal impact". 

@mfeurer can probably tell you more but previous more comprehensive attempts at ensembling suffered from high amounts of over-fitting. With respect to this implementation,
the only time I could see the case in which the last ensemble is not the best is when the ensemble has been highly over-fitted to the data and if we ever reach this case, we've already overfit horribly and optimizing the last bit of the score would not help that.

Hope this helps answer this question :)

Best,
Eddie

**Comment by mfeurer** [comment]
Hi,

This is a correct observation @LennartPurucker and we should fix this. I don't think this will have a large impact and I don't recall if we have specifically investigated this, maybe @franchuterivera remembers.

Have you found any other implementation differences that might be worth adapting? I know that Auto-Gluon uses a 2nd metric for tie breaking instead of the random tie breaking we use, but it would be good to know if you found something else.

**Comment by LennartPurucker** [comment]
Hey, 

Thank you for your quick responses!

@eddiebergman, I think your intuition is correct in regard to the benefits of step 4 on average. Furthermore, employing any kind of early stopping reduces the algorithm's already small 'sample size'. However, in certain use cases, additional protection against overfitting could still be useful 'enough' in my opinion. In general, avoiding overfitting is the biggest problem and the reason for the success of the Caruana et al. ensemble selection.  

I took the time today to perform a simple empirical evaluation of this problem. Please be aware that the setting for the evaluation (which base models were used; which metric optimized; which datasets selected) is from an AutoML-related use case and is potentially not generalizable to any application. Moreover, I can not yet share all details on it publicly (until soon at a certain conference :).  In essence, I tried to simulate auto-sklearn and its default value “ensemble_nbest=50” by taking 31 datasets with ca. 50 top-performing base models per dataset. Moreover, I took the implementation of ensemble selection from auto-sklearn and extended it with code to select the best ensemble at the end (like the “use_best” of AutoGluon). I evaluated both using 10-fold cross-validation whereby each base model was trained on the training data of a fold and the base models’ predictions on the fold’s validation data were split (frac=0.5) into data for ensemble selection and evaluation. 

My first observation is that, for all datasets, the final ensemble size (the number of selected base models) was smaller than 50. In other words, the best ensemble on the validation data was not once equal to the last ensemble once. However, in a head-to-head comparison, Ensemble Selection without use_best (ES) won 17 times (~55\%) and ensemble selection with use_best (ES-UB) won 14 times (~45\%). The average ROC AUC score was 0.7437 for ES and 0.7422 for ES-UB. Looking at the results per dataset, it seems that for larger datasets the effect is almost fully negligible and the difference in performance starts only at the fourth decimal point. This could also be the result of floating point errors. The opposite seems to happen for small datasets. Here, a difference in performance is already visible in the second decimal. 

These results should only be used to boost our intuition. I do not think they are fully representative, because the data used for ensemble selection became very small very quickly with 10-fold and a split fraction of 0.5. I am planning on doing a better evaluation with larger validation datasets and inner cross-validation in the future. 

--- 

@mfeurer, I am not sure whether the differences are worth adapting, but I can share my thoughts on existing differences:

- **Sorted Initialization:** Auto-sklearn does not have a 'sorted initialization' of the original paper. AutoGluon has implemented it, but commented out. I think this is not worth adapting because it was mainly advantageous due to it sorting and using a subset of the 2000 base model based on performance. This is achieved implicitly by Auto-sklearn and AutoGluon, since only the top-n models found during the model selection phase are used as base models at all. 
- **Tie-breaking** Indeed, AutoGluon has a second (fixed) metric for tie-breaking instead of random choice. Employing ensemble selection with a different metric than the metric that is to be optimized can also produce competitive performance, see [Caruana et al. (2006)]( https://ieeexplore.ieee.org/document/4053111). Hence, identifying that in a tie, ensemble A performs worse than Ensemble B for another metric could still make a positive difference for the generalized performance of the metric that is to be optimized. 
- **Time Limit**: AutoGluon supports a time limit for ensemble selection. This works well with ensemble selection, because the algorithm can stop after each iteration. Moreover, this can enable ensemble selection to be stopped by time instead of iterations. The time limit could be useful in scenarios where we do not have enough time to train another model, but enough time to do ensemble selection with existing models until the limit is reached. 
- **Early Stopping Based on Epsilon**: The re-implementation of the original paper supports the option to stop the iterations early if the performance does not change more than a constant epsilon. This is an alternative, more efficient, option to selecting the best models as a final step. But, since ensemble selection with only a few models (e.g., N=50) is already very efficient depending on the metric, it might not be worth stopping early and missing a potential increase in ensemble performance in later iterations. 
- **Rounding Scores**: AutoGluon can round the scores/losses to avoid a base model being selected due to floating-point errors. The floating-point errors can make two models that have identical predictions both be selected in subsequent iterations. Hence, rounding is useful to reduce the number of selected base models and thus decrease the inference time. 
- **Bagging**: All implementations have a parameter for bagging, but only the re-implementation has it fully implemented/supported. I don't think that AutoML tools require the use of bagging yet. The original paper by Caruana et al. (2004) used it to reduce overfitting in the face of thousands of models. Moreover, in Caruana et al. (2006), the results indicate that bagging becomes less useful when models are pre-pruned (by performance) and fewer base models to select from exist. Both Auto-sklearn and AutoGluon only keep a small (N<100)  set of models which are pre-pruned by performance by default.
- **Prediction Data**: Based on the implementation of AutoGluon, its ensemble selection algorithm could receive out-of-fold predictions generated by n-repeated k-fold bagging. These would be much less prone to overfitting by ensemble selection than Auto-sklearn’s out-of-fold prediction from cross- or holdout-validation. I do not think that Auto-sklearn should also support n-repeated k-fold bagging but it is important to keep this difference in mind when comparing the potential gain of ensemble selection in these two frameworks. 

These are all differences that I am aware of. I hope it helps. 

Best Regards,
Lennart

**Comment by mfeurer** [comment]
Hi @LennartPurucker 

Thanks for your explanation.

> These results should only be used to boost our intuition. I do not think they are fully representative, because the data used for ensemble selection became very small very quickly with 10-fold and a split fraction of 0.5. I am planning on doing a better evaluation with larger validation datasets and inner cross-validation in the future.

Do you have any timeline on this?

> Sorted Initialization

We tried this but didn't find any improvements in the AutoML setting. My intuition is similar to yours and that this makes most sense when you have 2000 base models as in the Caruana paper and not when you have a varying number as in AutoML.

> Tie-breaking

We should try that out. @franchuterivera  did some tests earlier on and found that using log_loss as a tie-breaker can help with auc.

> time limit

We could pass the time limit to the ensemble creator. Then, when the ensemble creator returns, we could check the validation score to see whether we use the new ensemble (because it might be worse since it might have only done a third of the iterations).

> Early stopping

This would add another hyperparameter which I strongly argue against.

> Rounding Scores

This sounds very useful to keep the ensemble tiny

> Bagging

We have never found this to work, maybe because of pruning, maybe because we have less than than the 2000 models Caruana et al. used.

> Prediction data

As you mentioned, this is a difference in the algorithm and we can't really do anything about this.

This leaves us with time limit, use best, tie-breaking and rounding scores that might be worth looking into. If you provide a PR/multiple PRs we can benchmark these changes on the AutoML benchmark datasets to see their impact. I'd leave out the time limit for now because I'd be worried that we need to adapt quite a bit of code in Auto-sklearn to cater for that.

**Comment by LennartPurucker** [comment]
Hi @mfeurer 

I am always happy to talk about ensembles! 


> Do you have any timeline on this?

Not really. Depending on the progress with my current work, a few months at best. 

> If you provide a PR/multiple PRs we can benchmark these changes on the AutoML benchmark datasets to see their impact

That would be good data as well. I will look into PRs for use best, tie-breaking, and rounding scores. 
I might also add the time limit to the code for ensemble selection because it is quite easy to do and could be used in the future. 


Regarding this issue, my original question was answered already. Should I close this issue or should I let it stay open as a PR reference?

**Comment by LennartPurucker** [comment]
Hey, @mfeurer @eddiebergman 

I have opened a PR for use best, tie-breaking, and rounding scores (losses).

I have not added anything related to a time limit, because I did not fully grasp what limit and how the limit could be passed to the ensemble selection. I agree now that this is better left on hold as @mfeurer said.
